core/src/main/scala/kafka/server/AbstractFetcherThread.scala

```diff
@@ -454,15 +492,23 @@ abstract class AbstractFetcherThread(name: String,
     * truncation completed if their offsetTruncationState indicates truncation completed
     *
     * @param fetchOffsets the partitions to update fetch offset and maybe mark truncation complete
+    * @param maySkipTruncation true if we can stay in Fetching mode and perform truncation later based on
+   *                           diverging epochs from fetch responses.
     */
-  private def updateFetchOffsetAndMaybeMarkTruncationComplete(fetchOffsets: Map[TopicPartition, OffsetTruncationState]): Unit = {
+  private def updateFetchOffsetAndMaybeMarkTruncationComplete(fetchOffsets: Map[TopicPartition, OffsetTruncationState],
+                                                              maySkipTruncation: Boolean): Unit = {
     val newStates: Map[TopicPartition, PartitionFetchState] = partitionStates.partitionStateMap.asScala
       .map { case (topicPartition, currentFetchState) =>
         val maybeTruncationComplete = fetchOffsets.get(topicPartition) match {
           case Some(offsetTruncationState) =>
-            val state = if (offsetTruncationState.truncationCompleted) Fetching else Truncating
+            val (state, lastFetchedEpoch) = if (offsetTruncationState.truncationCompleted)
+              (Fetching, latestEpoch(topicPartition))
+            else if (maySkipTruncation && currentFetchState.lastFetchedEpoch.nonEmpty)
+              (Fetching, currentFetchState.lastFetchedEpoch)
```

Comment:
I'm a little uncertain about this case. If we have truncated to an earlier offset, wouldn't we also need to reset last fetched epoch? I am thinking we should remove this check and modify the first one:
```scala
val (state, lastFetchedEpoch) = if (maySkipTruncation || offsetTruncationState.truncationCompleted)
  (Fetching, latestEpoch(topicPartition))
```
I might be missing something though.

Reply:
I was trying to to optimize for the case where lastFetchEpoch didn't need to be reset, but you are right, it would need resetting after truncation. Updated to use `latestEpoch(topicPartition))`.